File: src/main/java/server/game/DeckBuilder.java
package server.game;

import proto.dto.Card;
import proto.dto.CardColor;
import proto.dto.CardType;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Класс утилиты для построения стандартной колоды UNO.
 * Создаёт все 108 карт в полной колоде UNO, используя энумы DTO.
 */
public class DeckBuilder {

    /**
     * Создаёт полную стандартную колоду UNO из 108 карт.     *
     */
    public static List<Card> createStandardDeck() {
        List<Card> deck = new ArrayList<>();
        CardColor[] colors = {CardColor.RED, CardColor.BLUE, CardColor.GREEN, CardColor.YELLOW};

        // Add number cards (0-9 for each color)
        for (CardColor color : colors) {
            // One 0 card per color
            deck.add(new Card(generateCardId(color, CardType.NUMBER, 0), color, CardType.NUMBER, 0));

            // Two of each number 1-9 per color
            for (int number = 1; number <= 9; number++) {
                deck.add(new Card(generateCardId(color, CardType.NUMBER, number), color, CardType.NUMBER, number));
                deck.add(new Card(generateCardId(color, CardType.NUMBER, number) + "_2", color, CardType.NUMBER, number));
            }

            // Two of each action card per color
            // Skip cards
            deck.add(new Card(generateCardId(color, CardType.SKIP, null), color, CardType.SKIP, null));
            deck.add(new Card(generateCardId(color, CardType.SKIP, null) + "_2", color, CardType.SKIP, null));

            // Reverse cards
            deck.add(new Card(generateCardId(color, CardType.REVERSE, null), color, CardType.REVERSE, null));
            deck.add(new Card(generateCardId(color, CardType.REVERSE, null) + "_2", color, CardType.REVERSE, null));

            // Draw Two cards
            deck.add(new Card(generateCardId(color, CardType.DRAW_TWO, null), color, CardType.DRAW_TWO, null));
            deck.add(new Card(generateCardId(color, CardType.DRAW_TWO, null) + "_2", color, CardType.DRAW_TWO, null));
        }

        // Add Wild cards (4 of each type)
        for (int i = 1; i <= 4; i++) {
            deck.add(new Card("WILD_" + i, CardColor.WILD, CardType.WILD, null));
            deck.add(new Card("WILD_DRAW_FOUR_" + i, CardColor.WILD, CardType.WILD_DRAW_FOUR, null));
        }

        // Shuffle the deck
        Collections.shuffle(deck);

        return deck;
    }

    private static String generateCardId(CardColor color, CardType type, Integer number) {
        StringBuilder id = new StringBuilder();
        id.append(color.name()).append("_");
        id.append(type.name());
        if (number != null) {
            id.append("_").append(number);
        }
        return id.toString();
    }

    public static DeckPiles createDeckPiles() {
        List<Card> fullDeck = createStandardDeck();

        // Draw pile starts with all cards except the top one
        List<Card> drawPile = new ArrayList<>(fullDeck.subList(1, fullDeck.size()));

        // Discard pile starts with the top card
        List<Card> discardPile = new ArrayList<>();
        discardPile.add(fullDeck.get(0));

        return new DeckPiles(drawPile, discardPile);
    }

    public static class DeckPiles {
        private final List<Card> drawPile;
        private final List<Card> discardPile;

        public DeckPiles(List<Card> drawPile, List<Card> discardPile) {
            this.drawPile = new ArrayList<>(drawPile);
            this.discardPile = new ArrayList<>(discardPile);
        }

        /**
         * Берите карту из стопки.
         * Если стопка для добора пуста, перемешайте стопку сброса (кроме верхней карты) в стопку.
         */
        public Card drawCard() {
            if (drawPile.isEmpty()) {
                reshuffleDiscardIntoDraw();
            }

            if (drawPile.isEmpty()) {
                throw new IllegalStateException("No cards available to draw");
            }

            return drawPile.remove(drawPile.size() - 1);
        }

        public void playCard(Card card) {
            discardPile.add(card);
        }

        /**
         * Возьмите текущую верхнюю карту из стопки сброса.
         */
        public Card getTopCard() {
            if (discardPile.isEmpty()) {
                throw new IllegalStateException("Discard pile is empty");
            }
            return discardPile.get(discardPile.size() - 1);
        }

        /**
         * Перемешайте все карты из стопки сброса, кроме верхней, в стопку для добора.
         */
        private void reshuffleDiscardIntoDraw() {
            if (discardPile.size() <= 1) {
                return; // Nothing to reshuffle
            }

            // Keep the top card in discard pile
            Card topCard = discardPile.remove(discardPile.size() - 1);

            // Move all other cards to draw pile and shuffle
            drawPile.addAll(discardPile);
            discardPile.clear();
            Collections.shuffle(drawPile);

            // Put the top card back
            discardPile.add(topCard);
        }
    }
}

File: src/main/java/server/game/DeckBuilder.kt
package server.game

import proto.dto.Card
import proto.dto.CardColor
import proto.dto.CardType
import java.util.*

object DeckBuilder {
    fun createStandardDeck(): MutableList<Card?> {
        val deck: MutableList<Card?> = ArrayList<Card?>()
        val colors = arrayOf(CardColor.RED, CardColor.BLUE, CardColor.GREEN, CardColor.YELLOW)

        colors.forEach { color ->
            deck.add(
                Card(
                    id = generateCardId(
                        color = color,
                        type = CardType.NUMBER,
                        number = 0
                    ),
                    color = color,
                    type = CardType.NUMBER,
                    number = 0
                )
            )

            repeat(9) { number ->
                deck.add(
                    Card(
                        id = generateCardId(
                            color = color,
                            type = CardType.NUMBER,
                            number = number
                        ),
                        color = color,
                        type = CardType.NUMBER,
                        number = number
                    )
                )

                deck.add(
                    Card(
                        id = generateCardId(
                            color = color,
                            type = CardType.NUMBER,
                            number = number
                        ) + "_2",
                        color = color,
                        type = CardType.NUMBER,
                        number = number
                    )
                )
            }

            deck.add(
                Card(
                    id = generateCardId(
                        color = color,
                        type = CardType.SKIP,
                        number = null
                    ),
                    color = color,
                    type = CardType.SKIP,
                    number = null
                )
            )
            deck.add(
                Card(
                    id = generateCardId(
                        color = color,
                        type = CardType.SKIP,
                        number = null
                    ) + "_2",
                    color = color,
                    type = CardType.SKIP,
                    number = null
                )
            )

            deck.add(
                Card(
                    id = generateCardId(
                        color = color,
                        type = CardType.REVERSE,
                        number = null
                    ),
                    color = color,
                    type = CardType.REVERSE,
                    number = null
                )
            )

            deck.add(
                Card(
                    id = generateCardId(
                        color = color,
                        type = CardType.REVERSE,
                        number = null
                    ) + "_2",
                    color = color,
                    type = CardType.REVERSE,
                    number = null
                )
            )

            deck.add(
                Card(
                    generateCardId(color, CardType.DRAW_TWO, null),
                    color,
                    CardType.DRAW_TWO,
                    null
                )
            )

            deck.add(
                Card(
                    id = generateCardId(
                        color = color,
                        type = CardType.DRAW_TWO,
                        number = null
                    ) + "_2",
                    color = color,
                    type = CardType.DRAW_TWO,
                    number = null
                )
            )
        }

        repeat(9) { index ->
            val number = index + 1
            deck.add(
                Card(
                    id = "WILD_$index",
                    color = CardColor.WILD,
                    type = CardType.WILD,
                    number = null
                )
            )

            deck.add(
                Card(
                    id = "WILD_DRAW_FOUR_$index",
                    color = CardColor.WILD,
                    type = CardType.WILD_DRAW_FOUR,
                    number = null
                )
            )
        }

        deck.shuffle()
        return deck
    }

    private fun generateCardId(
        color: CardColor,
        type: CardType,
        number: Int?
    ): String {
        val id = StringBuilder()
        id.append(color.name).append("_")
        id.append(type.name)
        if (number != null) {
            id.append("_").append(number)
        }
        return id.toString()
    }

    fun createDeckPiles(): DeckPiles {
        val fullDeck = createStandardDeck()

        val drawPile = ArrayList(fullDeck.subList(1, fullDeck.size))

        val discardPile = ArrayList<Card?>()
        discardPile.add(fullDeck[0])

        return DeckPiles(drawPile, discardPile)
    }

    class DeckPiles(
        private val drawPile: MutableList<Card?>,
        private val discardPile: MutableList<Card?>
    ) {

        val topCard: Card
            get() {
                check(!discardPile.isEmpty()) { "Discard pile is empty" }
                return discardPile[discardPile.size - 1]!!
            }

        fun drawCard(): Card? {
            if (drawPile.isEmpty()) {
                reshuffleDiscardIntoDraw()
            }

            check(!drawPile.isEmpty()) { "No cards available to draw" }

            return drawPile.removeAt(drawPile.size - 1)
        }

        fun setTopCard(card: Card) {
            discardPile[discardPile.size - 1] = card
        }

        fun playCard(card: Card?) {
            discardPile.add(card)
        }

        private fun reshuffleDiscardIntoDraw() {
            if (discardPile.size <= 1) {
                return
            }

            val topCard = discardPile.removeAt(discardPile.size - 1)

            drawPile.addAll(discardPile)
            discardPile.clear()
            drawPile.shuffle()

            discardPile.add(topCard)
        }
    }
}